chinese/color.py

```diff
@@ -60,14 +60,16 @@ def _repl(p):
             pattern = d[target]
             text = ''
             for syllable in word.split():
-                if search(f'^{pattern}$', syllable):
-                    text += sub(f'^{pattern}$', _repl, syllable, IGNORECASE)
+                if search(f'^{pattern}$', syllable, flags=IGNORECASE):
+                    text += sub(f'^{pattern}$', _repl, syllable, flags=IGNORECASE)
                 elif has_ruby(syllable):
                     if ruby_whole:
                         pattern = RUBY_REGEX
                     else:
                         pattern = HALF_RUBY_REGEX
-                    text += sub(pattern, _repl, syllable, IGNORECASE)
+                    # For some reason HALF_RUBY_REGEX is case-insensitive per se, and RUBY_REGEX is case-sensitive
+                    # And for ruby we do not need to keep uppercase
+                    text += sub(pattern, _repl, syllable, flags=IGNORECASE).lower()
```

Comment:
Hrm, should we make `RUBY_REGEX` also match on `A-Z`? Looking around the code base I don't actually see anywhere that uses this `ruby_whole` branch except for tests... what was supposed to be the use, can we rip it out? I don't use ruby so I don't have the context to know what was originally trying to be done here.

I am a little dubious of the `.lower()` there, like I said I don't use ruby but wouldn't this result in people not being abel to specify pinyin with their own case? I think it someone wanted to not have uppercase they should just not use uppercase rather than have this choice forced on them.

Reply:
Wow, thanks for reviewing!

1) I tried to look into these regexps, but it's all very confusing because it was edited by different people. Some are inconsistent. In an ideal word all these patterns should be refactored, but... At least for now case sensitivity is handled by the IGNORECASE flag.

2) My rationale for forcing lowercase in ruby is as follows.

We use pinyin like a transcription to get the readings. And as a reading aid (like say hiragana). In the latter case we sometimes have complete sentences in pinyin written western-style with capitalization and so on (for example in lower HSK, children's books). It was exactly my case - I've got an HSK deck with sentences with pre-filled pinyin, and boom, all sentences were marked with the 5th tone.

But for ruby, I believe, it is used strictly as a transcription (like zhuyin/hiragana), never as a complete sentence. So it is only logical to convert it to lowercase to make it consistent and improve readability. I haven't used ruby that much too but decided to give it a try on the back of the cards to speed up reviewing.


So, TL;DR - I don't think that somebody will ever want ruby with uppercase letters. But your proposal with the config looks good and not very difficult.